Fusion de zds-antispam dans zds-site

Jan-Ha-He opened the pull request:
## Description

Dans cette pull request, nous fusionnons le module **zds-antispam** dans **zds-site** afin de supprimer la dépendance aux requêtes HTTP et permettre un accès direct aux données via l’ORM de Django. Concrètement, cela permettra :

- De simplifier l’architecture en évitant des appels à l’API HTTP ;
- Un accès plus facile aux biographies des utilisateurs;

### Fichiers modifiés

- **`signals.py`**  
Exécute le Spam Detector à chaque mise à jour d'une biographie.

- **`spam_detector.py`**  
Contient la logique principale du module antispam : analyse des contenus et règles permettant de marquer un texte comme suspect.

- **`spam_training.py`**  
Regroupe les fonctions nécessaires à l’entraînement ou à la mise à jour des règles de détection de spam, en utilisant les données recueillies sur Zeste de Savoir.

---

## Contrôle Qualité

Pour vérifier la bonne intégration de zds-antispam :

1. Lancez `python manage.py migrate` puis `yarn test` pour vous assurer du bon fonctionnement global.  
2. Créez ou modifiez un profil utilisateur avec une biographie suspecte.  
3. Vérifiez que le contenu problématique est bien signalé par le nouveau module antispam (consultez les logs ou l’interface d’administration).  

En cas de problème ou de question, n’hésitez pas à le mentionner dans cette discussion.

_TODO:_

- [x] Création de tests
- [x] Intégration de données de formation
- [x] Mise en œuvre du rythme de formation
- [x] Assurer la persistance du modèle entraîné

Commit:
Ajoute les liens pour "Sujets créés", "Sujets suivis", "Messages postés" dans la sidebar des forums.
Fix #6467

- Vérifiez que les liens s'affichent bien dans la sidebar.
- Testez que les liens redirigent vers les bonnes pages.
- Assurez-vous que le comportement est correct sur mobile et desktop.

wassimaarab committed:
Merge branch 'zestedesavoir:dev' into dev

Jan-Ha-He committed:
Fusion de zds-antispam dans zds-site

Jan-Ha-He committed:
Fusion de zds-antispam dans zds-site

Situphen commented on zds/utils/spam_training.py:
```diff
@@ -0,0 +1,74 @@
+_bio = []
+_can_read = []
+
+current_dir = os.path.dirname(os.path.abspath(__file__))
+json_path = os.path.join(current_dir, "spamdata.json")
```
Dans les projets Django, on préfère en règle générale créer les chemins à partir de `BASE_DIR` pour partir de la même base partout. De plus, on préfère travailler avec des objets `Path` que le module `os.path` quand c'est possible. La documentation donne les équivalences entre les deux modules : https://docs.python.org/fr/3.13/library/pathlib.html#corresponding-tools
```py
from django.conf import settings
chemin = settings.BASE_DIR / "sous-dossier" / "fichier.txt"
```

Situphen commented on zds/utils/signals.py:
```diff
@@ -1,7 +1,63 @@
+import logging
+import os
+
+# Logger Setup
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.ERROR)
+current_dir = os.path.dirname(os.path.abspath(__file__))
+log_file = os.path.join(current_dir, "spam_signals.log")
+
+# File Handler
+handler = logging.FileHandler(log_file)
+handler.setLevel(logging.ERROR)
+formatter = logging.Formatter("%(asctime)s - %(levelname)s - %(message)s")
+handler.setFormatter(formatter)
```
Dans les fichiers de code du projet, on définit le *logger* avec seulement cette simple ligne : 

```suggestion
logger = logging.getLogger(__name__)
```

Le reste est géré dans la configuration : 

https://github.com/zestedesavoir/zds-site/blob/5b98a51b5aece0acdfa15d65b0efc71577dddba9/zds/settings/abstract_base/django.py#L253-L287

Situphen commented on zds/utils/spam_detector.py:
```diff
@@ -0,0 +1,96 @@
+        for user in self.reported_users:
+            self.logger.info(f"User {user} already reported")
+
+        current_dir = os.path.dirname(os.path.abspath(__file__))
+        log_file = os.path.join(current_dir, "spam_detector.log")
+
+        handler = logging.FileHandler(log_file, mode="a")
+        formatter = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
+        handler.setFormatter(formatter)
+        self.logger.addHandler(handler)
```
Idem, géré dans la configuration du projet
```suggestion
```

Situphen commented on zds/utils/spam_detector.py:
```diff
@@ -0,0 +1,96 @@
+    reported_users = []
+
+    def __init__(self):
+        self.logger = logging.getLogger("zds.spam")
+        self.logger.setLevel(logging.ERROR)
```
À remplacer par `logger = logging.getLogger(__name__)` juste après les lignes d'importation
```suggestion
```

Situphen commented on zds/utils/spam_detector.py:
```diff
@@ -0,0 +1,96 @@
+from django.contrib.auth.models import User
+
+
+class SpamDetector:
```
Je pense que le code gagnerait en lisibilité en remplaçant cette classe par une fonction, car le code est assez linéaire.

Situphen commented on zds/utils/spam_detector.py:
```diff
@@ -0,0 +1,96 @@
+
+class SpamDetector:
+
+    reported_users_file = "reported_users.txt"
+    reported_users = []
```
Dans le fonctionnement actuel, le code est exécuté toutes les 5 minutes donc ce fichier permet de ne pas vérifier tout le temps la même biographie. Étant donné que dans le nouveau fonctionnement ce code est exécuté pour une biographie seulement à sa création ou à sa modification, ce fichier n'a plus d'utilité ! Tu peux donc retirer tout ce qui est lié à `reported_users`.

Situphen reviewed (changes requested):
Merci pour cette PR ! Voici quelques remarques qui me viennent après une première lecture, n'hésites pas à argumenter si tu as un avis différent ;)

Remarques concernant tous les fichiers : 
- Pour une meilleure lisibilité, il est recommandé de grouper les lignes d'importation (dans l'ordre : modules standard de Python, puis modules installés avec Pip, puis modules de ton projet) et de les trier par ordre alphabétique. Ça peut être fait automatiquement avec `isort` et je viens de créer une PR pour l'inclure dans le projet (#6721).
- Tu es parti du code existant et c'est très bien, mais n'hésites pas à restructurer complètement le code (passer d'une classe à une fonction, renommer les variables, etc.) car d'une part le cas d'usage est très différent (on n'est plus du tout sur un script qui tourne tous les X minutes et fait des appels API) et d'autre part le code est vieux !

Jan-Ha-He committed:
Merge pull request #1 from zestedesavoir/dev

Mise a jour du fork

Jan-Ha-He committed:
Merge pull request #2 from wassimaarab/dev

Merge tests pour zds-antispam integration

Jan-Ha-He committed:
Ajout de la génération de profils spam pour tester le filtre.

Jan-Ha-He committed:
Merge remote-tracking branch 'origin/dev' into dev

Commit:
Suppression de la création d'utilisateurs en base de données
Modification spam_training.py et spam_detector.py :
	1-Remplacement des données dynamiques par un jeu de données statique
	2-Utilisation de données factices pour l'entraînement du modèle
Ajout des tests unitaires qui couvrent :
	1-Profils sans biographie
     	2-Contenu spam
     	3-Contenu valide

Lancer python3 manage.py test zds.utils.tests.tests_antispam

Commit:
Suppression de la création d'utilisateurs en base de données
Modification spam_training.py et spam_detector.py :
	1-Remplacement des données dynamiques par un jeu de données statique
	2-Utilisation de données factices pour l'entraînement du modèle
Ajout des tests unitaires qui couvrent :
	1-Profils sans biographie
     	2-Contenu spam
     	3-Contenu valide

Lancer python3 manage.py test zds.utils.tests.tests_antispam

wassimaarab reviewed (commented):
la commit 717cb10 est une duplication involontaire due à un conflit de merge.
Merci d’ignorer celle-ci et de vous référer à la commit précédente 385b8f5 pour les vrais changements

Jan-Ha-He committed:
Reorganisation dans un nouvel package zds.antispam.

Jan-Ha-He committed:
Ajout d'exigences, amélioration de la structure du code, ajout de la persistance et de retrain_spam_filter.py pour les tâches planifiées.

Jan-Ha-He committed:
Ajout d'exigences, amélioration de la structure du code, ajout de la persistance et de retrain_spam_filter.py pour les tâches planifiées.

Jan-Ha-He committed:
Merge pull request #3 from Jan-Ha-He/restructuring

Restructuring

Jan-Ha-He committed:
Ajout de reentraîner le modèle que nécessaire.

Jan-Ha-He committed:
Merge pull request #4 from Jan-Ha-He/restructuring

Ajout de reentraîner le modèle que nécessaire

philippemilink commented on zds/antispam/retrain_spam_filter.py:
Il faudrait plutôt créer une [commande Django](https://docs.djangoproject.com/en/5.1/howto/custom-management-commands/), qu'on pourrait par exemple appeler avec `python3 manage.py antispam_train`. Voir par exemple la commande qui gère l'indexation dans le moteur de recherche : `zds/search/management/commands/search_engine_manager.py`.

Jan-Ha-He replied:
Fixed

philippemilink commented on zds/antispam/spam_detector.py:
```diff
@@ -0,0 +1,56 @@
+class SpamDetector:
+    def __init__(self, model_file="spam_filter_model.pkl"):
+        self.logger = logging.getLogger(__name__)
+        self.logger.setLevel(logging.ERROR)
```
```suggestion
```
A priori c'est inutile, cf le commentaire de Situphen https://github.com/zestedesavoir/zds-site/pull/6720#discussion_r1988152435

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/antispam/spam_detector.py:
```diff
@@ -0,0 +1,56 @@
+        try:
+            profile = Profile.objects.get(user__username=username)
+
+            message = _("Spam")
```
```suggestion
            message = _("Spam potentiel")
```
(comme ce qui est fait actuellement)

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/antispam/spam_detector.py:
```diff
@@ -0,0 +1,56 @@
+            message = _("Spam")
+
+            Alert.objects.create(
+                author=User.objects.get(username="bot"),
```
[On avait dit que ce serait bien de garder un utilisateur dédié à l'antispam](https://zestedesavoir.com/forums/sujet/11108/reunion-des-dev-de-zeste-de-savoir/?page=9#fusion-zds-antispam-dans-zds-site).

Sans doute qu'il faut donc ajouter les paramètres nécessaires pour pouvoir définir quel utilisateur utiliser pour rapporter les suspicions de spam, comme c'est fait pour le compte `bot` (cherchez `anonymous_account` dans le code).

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/antispam/spam_detector.py:
```diff
@@ -0,0 +1,56 @@
+        else:
+            self.logger.info("✔️  %s's biography doesn't look like spam" % profile.user.username)
+
+        self.logger.info(f"Profile checked: {profile.user.username}\n" f"Biography: {profile.biography}\n" f"{'='*50}")
```
```suggestion
```
Ce message de log semble redondant : que la biographie ressemble à du spam ou non, on affiche déjà un message de log.

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/antispam/spam_detector.py:
```diff
@@ -0,0 +1,56 @@
+
+        except Exception as e:
+            self.logger.error(f"Error on creation of spam report for {username}: {str(e)}")
+            return False
```
La valeur de retour de cette fonction n'est pas utilisée.

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/antispam/spam_model_manager.py:
```diff
@@ -0,0 +1,97 @@
+import logging
+import os
+
+import joblib
+from factory.fuzzy import FuzzyText
```
```suggestion
from factory.fuzzy import FuzzyText
import joblib
```

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/antispam/spam_model_manager.py:
```diff
@@ -0,0 +1,97 @@
+        self.count_vect = None
+        self.tfidf_transformer = None
+        self.logger = logging.getLogger(__name__)
+        self.logger.setLevel(logging.ERROR)
```
```suggestion
```

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/antispam/spam_model_manager.py:
```diff
@@ -0,0 +1,97 @@
+
+
+class SpamModelManager:
+    def __init__(self, model_file="spam_filter_model.pkl"):
```
Dans l'idéal, il faudrait que le chemin et le nom du fichier soit un paramètre de l'application. Je pense que vous pouvez vous inspirer de ce qui est fait pour les fichiers qui donnent l'origine géographique des adresses IP : 
https://github.com/zestedesavoir/zds-site/blob/34555192ab615e431b0bd9bcea34b0173d7be834/zds/settings/abstract_base/zds.py#L12-L13

en le plaçant par défaut dans `BASE_DIR / `antispam-data` par exemple (ne pas oublier d'ajouter ce dossier au `.gitignore`)

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/antispam/spam_model_manager.py:
```diff
@@ -0,0 +1,97 @@
+
+        profiles = Profile.objects.all()
+
+        if profiles.count() < 100:
+            # Generate fake data for testing
+            spam_profiles = [f"spam {FuzzyText(prefix='spammy').fuzz()} buy now free !!" for _ in range(50)]
+            non_spam_profiles = [f"correct {FuzzyText(prefix='normal').fuzz()} about ..." for _ in range(50)]
+
+            bios = spam_profiles + non_spam_profiles
+            can_read = [0] * len(spam_profiles) + [1] * len(non_spam_profiles)
```
Ça devrait être fait directement dans les tests (en production, on a déjà plus de 100 membres).

Jan-Ha-He replied:
Il y a un petit problème avec ça.
Dans load_fixtures.py, pendant la création des profils, on a déjà besoin du modèle entraîné pour passer les tests.
Mais comme le modèle est entraîné à partir des profils, cela crée un blocage.
Pour le résoudre, nous avons intégré la génération de données factices (dummy) dans le code.

philippemilink commented on zds/antispam/spam_model_manager.py:
```diff
@@ -0,0 +1,97 @@
+
+        return bios, can_read
+
+    def retrain(self):
```
Pourquoi pas juste `train()` ?

Jan-Ha-He replied:
Bonne remarque, c'est changé.

philippemilink commented on zds/antispam/spam_detector.py:
```diff
@@ -0,0 +1,56 @@
+
+        if self.model_manager.predict([biography])[0] == 0:
+            self.logger.info("✘  %s's biography looks like spam" % profile.user.username)
+            self.send_alert(None, profile.user.username)
```
```suggestion
            self.send_alert(None, profile.user.username)
            return True
```

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/antispam/spam_detector.py:
```diff
@@ -0,0 +1,56 @@
+            self.logger.info("✘  %s's biography looks like spam" % profile.user.username)
+            self.send_alert(None, profile.user.username)
+        else:
+            self.logger.info("✔️  %s's biography doesn't look like spam" % profile.user.username)
```
```suggestion
            self.logger.info("✔️  %s's biography doesn't look like spam" % profile.user.username)
            return False
```

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/antispam/tests/tests_spam_detector.py:
```diff
@@ -0,0 +1,38 @@
+        with patch.object(self.spam_detector, "send_alert") as mock_send_alert:
+            self.spam_detector.check_profile(self.profile3)
+            mock_predict.assert_called_once_with(["I love programming and open-source."])
+            mock_send_alert.assert_not_called()
```
C'est bizarre comme façon de tester (tout du moins ce n'est pas vraiment ce qu'on fait dans zds-site). On utilise `patch` quand on veut connaître l'état interne d'un objet, ou un mécanisme difficilement observable de l'extérieur.

Là il suffirait de faire : 
```py
self.assertTrue(self.spam_detector_check_profile(self.profile))
self.assertEqual(1, Alert.objects.count())
```
(la base de données est remise à zéro avant d'exécuter chaque méthode de test)

philippemilink commented on zds/antispam/tests/tests_spam_manager.py:
Idem, c'est sans doute plus simple de faire tous ces tests sans passer par des `@patch()` ...

philippemilink commented on zds/utils/management/commands/load_fixtures.py:
```diff
@@ -35,13 +35,13 @@
     """
-    nb_users = size * 10
-    cli.stdout.write(f"Nombres de membres à créer : {nb_users}")
+    nb_users = size * 50
```
```suggestion
    nb_users = size * 20
```

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/utils/management/commands/load_fixtures.py:
```diff
@@ -35,13 +35,13 @@
-    cli.stdout.write(f"Nombres de membres à créer : {nb_users}")
+    nb_users = size * 50
+    spam_ratio = 0.2  # 20% of profiles will be spam
+    cli.stdout.write(f"Nombres de membres à créer : {nb_users} (including {int(nb_users * spam_ratio)} spammers)")
```
```suggestion
    cli.stdout.write(f"Nombres de membres à créer : {nb_users} (dont {int(nb_users * spam_ratio)} spammeurs)")
```
(pour rester cohérent)

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/utils/signals.py:
```diff
@@ -1,7 +1,38 @@
+from django.dispatch import Signal, receiver
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.ERROR)
```
```suggestion
```

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/utils/signals.py:
```diff
@@ -1,7 +1,38 @@
+            return
+
+        logger.info(f"Spam check started for {instance.user.username}")
+        detector = SpamDetector()  # Uses the refactored SpamDetector
```
Selon moi le commentaire est inutile...

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/utils/signals.py:
Les apports à ce fichier devraient plutôt aller dans un fichier `zds/antispam/receivers.py` (regarder les autres fichiers `receivers.py` dans le dépôt). Ce fichier `zds/utils/signals.py` définit uniquement des signaux.

Jan-Ha-He replied:
Fixed

philippemilink commented on zds/utils/signals.py:
```diff
@@ -1,7 +1,38 @@
+    return apps.get_model("member", "Profile")
+
+
+@receiver(post_save)
```
Pour éviter d'avoir à faire `if sender == get_profile_model()`, vous pouvez faire : 
```py
@receiver(post_save, sender=Profile)
```

Jan-Ha-He replied:
Fixed

philippemilink reviewed (changes requested):
Pas mal de petits points à revoir.

Ce serait bien d'ajouter quelques commentaires dans le code pour expliquer le fonctionnement général de l'antispam. Et également rédiger la documentation correspondante.

Par contre, la façon dont vous avez conçu le code ne prend pas du tout en compte l'évolution qui permettrait de [chercher du spam dans d'autres champs](https://zestedesavoir.com/forums/sujet/11108/reunion-des-dev-de-zeste-de-savoir/?page=8#projet-%C3%A9tudiant). Mais il ne vous reste sans doute pas assez de temps pour mettre ça en place...

wassimaarab commented:
Pour réaliser la tache pour l'évolution, nous pensons qu’il suffit d’entraîner le modèle séparément sur chacun des champs (commentaire, biographie et contenu), puis de le lancer pour l’analyse de spam. Ensuite, on pourrait simplement adapter le message d’alerte en fonction du champ analysé. Est-ce bien la démarche à suivre, ou y a-t-il une autre approche recommandée ?

philippemilink commented:
> Pour réaliser la tache pour l'évolution, nous pensons qu’il suffit d’entraîner le modèle séparément sur chacun des champs (commentaire, biographie et contenu), puis de le lancer pour l’analyse de spam. Ensuite, on pourrait simplement adapter le message d’alerte en fonction du champ analysé. Est-ce bien la démarche à suivre, ou y a-t-il une autre approche recommandée ?

C'est pas tant d'appliquer le même processus qui est difficile, mais de coder la chose de façon à ce qu'on puisse très facilement ajouter les champs dans lesquels on souhaite chercher du spam. Dans l'idéal, pour dire au système qu'on souhaite chercher du spam dans un champ supplémentaire, ce serait juste une ligne de code ajouter. On en avait discuté [lors d'une réunion de dev's](https://zestedesavoir.com/forums/sujet/11108/reunion-des-dev-de-zeste-de-savoir/?page=8#projet-%C3%A9tudiant).

Jan-Ha-He committed:
Résolution de 14/22 des petits points (modifications des loggers, etc.)

Jan-Ha-He committed:
Résolution de 14/22 des petits points (modifications des loggers, etc.)

Jan-Ha-He committed:
Ajout de la commande antispam_train et déplacement des receivers vers antispam.

Jan-Ha-He committed:
Ajout de la commande antispam_train et déplacement des receivers vers antispam.

Jan-Ha-He committed:
Ajout d'un utilisateur antispam pour les notifications antispam.

Jan-Ha-He committed:
Ajout d'un utilisateur antispam pour les notifications antispam.

Jan-Ha-He committed:
Ajout du code pour créer et utiliser un nouvel dossier pour les fichiers antispam.

Jan-Ha-He committed:
Ajout du code pour créer et utiliser un nouvel dossier pour les fichiers antispam.

Jan-Ha-He committed:
Fusion de retrain() et train().

Jan-Ha-He committed:
Fusion de retrain() et train().

Jan-Ha-He committed:
Correction pour le dernier commit : il manquait load_model dans spam_model_manager.py.

Jan-Ha-He committed:
Correction pour le dernier commit : il manquait load_model dans spam_model_manager.py.

Jan-Ha-He committed:
Correction #2 pour le dernier commit : predict() manque aussi

Jan-Ha-He committed:
Correction #2 pour le dernier commit : predict() manque aussi

Jan-Ha-He committed:
Correction #3 pour le dernier commit : train(), pas retrain()

Jan-Ha-He committed:
Correction #3 pour le dernier commit : train(), pas retrain()

Jan-Ha-He committed:
Fixation éventuelle pour le chemin d'accès au fichier.

Jan-Ha-He committed:
Fixation éventuelle pour le chemin d'accès au fichier.

Jan-Ha-He committed:
Fixation #2 éventuelle pour le chemin d'accès au fichier.

Jan-Ha-He committed:
Fixation #2 éventuelle pour le chemin d'accès au fichier.

Jan-Ha-He committed:
Ajout de la possibilité de tester d’autres attributs pour le spam, avec le même modèle pour tous les attributs.

Jan-Ha-He committed:
Ajout de la possibilité de tester d’autres attributs pour le spam, avec le même modèle pour tous les attributs.

Jan-Ha-He committed:
Ajout de la possibilité d’avoir plusieurs modèles pour des attributs différents.

Jan-Ha-He committed:
Ajout de la possibilité d’avoir plusieurs modèles pour des attributs différents.

philippemilink commented on zds/antispam/apps.py:
Compare avec ce qui est fait dans `zds/notification/apps.py` ou `zds/featured/apps.py`, il y a quelques lignes dont on n'a pas besoin.

Jan-Ha-He replied:
Fixed

philippemilink commented on zds/antispam/receivers.py:
```diff
@@ -0,0 +1,20 @@
+    for field_config in spam_fields:
+        if isinstance(instance, field_config["model"]):
+            field_value = getattr(instance, field_config["field"], None)
+            if field_value:
```
Selon moi cette condition n'est pas nécessaire : si on ajoute un champ à vérifier à la liste `spam_fields`, alors il faut que l'objet décrivant le champ ait un attribut `field`. Si ce n'est pas le cas, c'est normal que ça se comporte mal.

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/antispam/receivers.py:
```diff
@@ -0,0 +1,20 @@
+                detector = SpamDetector()
+                # Pass the content_type (scope) to check_text
+                if detector.check_text(field_value, field_config["scope"]):
+                    detector.send_alert(instance, field_config["field"])
```
J'ajouterais  un `break` à la fin du `if isinstance(...)` : ce n'est pas la peine de continuer la boucle si on vient de trouver un élément de la liste qui correspond à cet objet.

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/antispam/spam_detector.py:
```diff
@@ -0,0 +1,81 @@
+                "scope": scope,
+                "text": _(f"Potential spam detected in {instance_info}, field '{field_name}'."),
+                "pubdate": datetime.now(),
+            }
```
```suggestion
            scope_to_alert_kwargs = {
                "PROFILE": "profile",
                "FORUM": "comment",
                "CONTENT": "content",
            }
            alert_kwargs = {
                "author": User.objects.get(username="antispam"),
                "scope": scope,
                "text": _(f"Potential spam detected in {instance_info}, field '{field_name}'."),
                "pubdate": datetime.now(),
                scope_to_alert_kwargs[scope]: instance,
            }
```
Mais peut-être qu'on peut définir le scope directement à la valeur attendue par `Alert` ? La question c'est : est-ce qu' on a envie/besoin de distinguer les messages sur le forum et les commentaires des contenus pour détecter du spam ?

Jan-Ha-He replied:
J’ai fait les modifications comme j’ai compris le commentaire, mais je ne suis pas 100 % sûr que c’est bien ce que le commentaire voulait dire.

Jan-Ha-He replied:
J'ai fait des changements comme j'ai compris le commentaire, je ne suis pas 100% sûre que c'était ce que était signifié par le commentaire

philippemilink commented on zds/antispam/spam_detector.py:
```diff
@@ -0,0 +1,81 @@
+            alert_kwargs = {
+                "author": User.objects.get(username="antispam"),
+                "scope": scope,
+                "text": _(f"Potential spam detected in {instance_info}, field '{field_name}'."),
```
On a tendance à écrire les messages en français.

Jan-Ha-He replied:
Fixed!

philippemilink commented on zds/antispam/spam_fields.py:
```diff
@@ -0,0 +1,17 @@
+        "model": Profile,
+        "field": "biography",
+        "scope": "PROFILE",
+        "get_instance_info": lambda instance: f"Profile of user '{instance.user.username}'",
```
J'ai l'impression que cet attribut correspond aux méthodes `__str__()` des des objets ?

Jan-Ha-He replied:
C’est changé pour le profil, mais pas pour les commentaires, car cela entraînerait du HTML en clair dans les alertes.

philippemilink commented on zds/antispam/spam_model_manager.py:
```diff
@@ -0,0 +1,88 @@
+            os.makedirs(directory)
+            self.logger.info(f"Created directory: {directory}")
+
+    def prepare_training_data(self, content_type):
+        """
+        Prepare training data for the given content type.
+        """
+        # Implement logic to fetch or generate training data based on content_type
+        bios = ["example spam text", "example non-spam text"]
+        labels = [0, 1]  # 0 for spam, 1 for non-spam
+        return bios, labels
+
```
Cette méthode n'a pas sa place dans cette classe, elle doit être directement dans les tests.

Jan-Ha-He replied:
En général, cette fonction doit aussi collecter des données pour l’entraînement de la base de données, mais il y avait une petite erreur dans le commit.
Pour la création des données de test, j’ai également ajouté un autre commentaire.

philippemilink commented on zds/antispam/spam_fields.py:
```diff
@@ -0,0 +1,17 @@
+spam_fields = [
+    {
+        "model": Profile,
+        "field": "biography",
```
On pourrait plutôt fournir une liste de champs : 
```py
"fields": ["biography", "sign"],
```

Jan-Ha-He replied:
Fixed!

Jan-Ha-He committed:
Correction de la génération des données de test et suppression des lignes inutiles dans apps.py.

Jan-Ha-He committed:
Correction de la génération des données de test et suppression des lignes inutiles dans apps.py.

Jan-Ha-He committed:
Correction de petits éléments + ajout du champ is_spam pour Profile et Comment

Jan-Ha-He committed:
Correction de petits éléments + ajout du champ is_spam pour Profile et Comment

Jan-Ha-He committed:
Changement de l’utilisation du scope dans send_alert.

Jan-Ha-He committed:
Changement de l’utilisation du scope dans send_alert.

Commit:
Tester le modele spam_detector.py

Vérification des tests :
	-Dans le répertoire racine, lance : python3 manage.py test zds.utils.tests.tests_antispam

Commit:
Tester le modele spam_manager.py

Vérification des tests :
	-Dans le répertoire racine, lance : python3 manage.py test zds.utils.tests.tests_spam_manager

Commit:
Ajouter la documentation pour le modèle antispam

Fichiers modifiés/ajoutés :

1-Modifié arborescence-back.rst : ajout le modèle antispam/ à l’arborescence de zds/

2-antispam.rst : rédaction de la documentation pour le modèle antispam

Jan-Ha-He committed:
Merge remote-tracking branch 'origin/dev' into dev

Jan-Ha-He committed:
Petites améliorations : structure des posts spam et journalisation des données d’entraînement

Situphen commented:
Bonjour,

Je me permets de repartir sur une autre branche (#6770) pour continuer votre travail !
